### ISSUE TITLE
Filter `Appointment` by values of `Status`

### ISSUE THREAD

**Issue by iaindillingham** [opensafely-core/ehrql#1307]
In [cohort-extractor][2], we filter [the `Appointment` table][1] by specific values of the `Status` column. According to [the short data report][5], appointments with these status values represented about 85% of all appointments, up to the middle of February 2023. We don't do the same in ehrQL.

Although it's not clear why we do this in cohort-extractor,[^1] I think it's reasonable to assume that we do this for good reasons. For consistency, we should do the same in ehrQL.

[^1]: [The documentation for `with_gp_consultations`][4] links to [a blog post from the King's Fund][3], which itself links to a report.

[1]: https://reports.opensafely.org/reports/opensafely-tpp-database-schema/#Appointment
[2]: https://github.com/opensafely-core/cohort-extractor/blob/fe1c9f84b896f6542db25e96ced4cf4681c75fe3/cohortextractor/tpp_backend.py#L2116-L2123
[3]: https://www.kingsfund.org.uk/blog/2016/05/crisis-general-practice
[4]: https://docs.opensafely.org/study-def-variables/#cohortextractor.patients.with_gp_consultations
[5]: https://jobs.opensafely.org/api/v2/releases/file/01GSZFT1B1P1N3SWC15C29897H

**Comment by evansd** [comment]
Yes, I agree. I think we need to make the raw, unfiltered data available one way or the other. Having a helper method for the filtered data seems like a sensible approach.

**Comment by iaindillingham** [comment]
Alternatively, could we create a new `gp_consultation` table?

My understanding is that in cohort-extractor, we wanted to expose only those rows of the `Appointment` table that represented GP consultations (the function is called `with_gp_consultations`, after all) and that the given values of the `Appointment.Status` column did so. The other values of the status column may represent other types of appointment, or be unreliable, or be unknown.

The reason for exposing this information now, in ehrQL, is for a study that's investigating GP consultations.

**Comment by sebbacon** [comment]
That's an interesting suggestion. It speaks to the vision I had ages ago around "contracts" as a place to negotiate (and iteratively develop) vendor-neutral concepts around primary care.


I note the [GP Connect FIHR documentation for Consultations](https://developer.nhs.uk/apis/gpconnect-1-5-0/accessrecord_structured_development_consultation_guidance.html) has a pretty good summary (at the top of that page) of the problems in this domain. Specifically, I think what we're talking about is an [Encounter](https://developer.nhs.uk/apis/gpconnect-1-5-0/accessrecord_structured_development_encounter.html) ("what can loosely be termed a ‘Date/Doctor/Place/Type’") and it looks like the only statuses allowed there are "unknown" and "finalised"; it then also has another column `type`, which can be any SNOMED code.

These valid statuses align with what Chris said in that Slack thread, I think. That is, there's "it happened" appointments; versus everything else.

I'm _not necessarily_ suggesting we do this, just thinking aloud, but _if_ we were to implement a new table, I suppose it would make sense to align it with the GP Connect one, which looks like mean we'd make the "raw" status available as SNOMED codes in a `type` field ([here's](https://termbrowser.nhs.uk/?perspective=full&conceptId1=305058001&edition=uk-edition&release=v20230412&server=https://termbrowser.nhs.uk/sct-browser-api/snomed&langRefset=999001261000000100,999000691000001104) the parent of Patient Encounter Status which includes _Cancelled_, _Did Not Attend_, etc), and _potentially_ add a helper column that reduces those to `finalised` and `unknown` statuses).

It seems to me our only choices right now are 
1. Expose just the raw data without any interpretation (i.e. views over the data) at all
2. Provide some interpretation (and minimally improve on the cohortextractor situation which is too opaque)

If we go for (1) we should be really careful to slather it in caveats as appointment data is simultaneous universally massively desired by the system, and hugely problematic.

Obviously the same is true for (2), but it possibly gives a clearer place to be somewhat opinionated and document those opinions along with the fact they're provisional (this also makes me think of that unused "vendor-specific notes" functionality)

I'd be interested to know if/how the team feels it's feasible to start trying out the idea of iterative-concept-development-through-opinionated-provisional-tables along these lines.

**Comment by iaindillingham** [comment]
Thanks Seb, it took me awhile to interpret your comment, and the linked documents, but what I think you're saying is that we could implement a new table that mapped `Appointment.Status` to SNOMED codes and add an extra property that reduced these SNOMED codes to two statuses: `"finished"` and `"unknown"` (as per `Encounter.status`). Researchers, we think, would be most interested in cases where `status == "finished"`.

Of the options you outline, we chose option 1.5 in cohort-extractor. We exposed most of the raw data -- 85% of all appointments -- and also slathered caveats over the documentation. We chose option 1 in ehrQL. Neither option seems satisfactory, and led to issues reconciling the two arms of the winter pressures project (which extracted data with ehrQL/SQL Runner and cohort-extractor respectively). For this reason, I'd been interested in trying iterative concept development. As well as the Manchester group, which @milanwiedemann is co-piloting, Ben mentioned that an LSHTM group are also interested in these data. So, it seems like a good time.

**Comment by Jongmassey** [comment]
FYI @milanwiedemann  https://github.com/opensafely-core/ehrql/pull/1482
